### PR TITLE
Evaluator: Allow making arbitrary file checks on the project's source, take 2

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits.

Note: The project's source tree will be cloned in the evaluator only if the rules needed it / if the file checks are actually used.

This implements parts of https://github.com/oss-review-toolkit/ort/issues/5621.

### CODE REVIEW

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt
Comment (sschuberth) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:332]: Commit message nits:
- "for dependency source" -> "for dependency source*s*" or "for the source code of dependencies" or similar
- "not have obvious need" -> "not have obvious need*s*" or just "is usually not required"
- "exection time" -> "exec*u*tion time"

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt
Comment (sschuberth) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:339]: Unrelated formatting change.

File: evaluator/src/main/kotlin/Evaluator.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/Evaluator.kt:43]: So, the whole script evaluation still gets a `SourceTreeResolver`... however, I was hoping to see a solution that does not change any host-side scripting classes at all, but only adds a new rule class derived from `OrtResultRule` which creates a `SourceTreeResolver` from the `OrtResult`.

Am I overlooking something why that's not possible? Do you want me to make a proposal based on your code?

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt
Comment (sschuberth) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:208]: I don't think we need this if the API was better, by moving the source code download to a dedicated rule.

File: evaluator/src/main/kotlin/Evaluator.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/Evaluator.kt:43]: > Am I overlooking something why that's not possible? Do you want me to make a proposal based on your code?



This code is using constructor injection of dependencies for testability. E.g. the test can inject an arbitrary directory as source code directory. I understand that you dislike constructor injection in this case. So, can you tell which way of injecting a directory from the test would be acceptable for you? (Ideally along with an argumentation why)

Having the above question answered would unblock the situation. Note that I understood from previous talks that you didn't like the construct, however I didn't get why and also not what alternative you want - which is why I didn't change it. It just didn't make sense to me.

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt
Comment (fviernau) [reply cli/src/main/kotlin/commands/EvaluatorCommand.kt:208]: I just had the idea of turning
```
fun ortResultRule(name: String, configure: OrtResultRule.() -> Unit) {
```
into 
```
fun ortResultRule(name: String, configure: OrtResultRule.() -> Unit, allowSourceCodeDownload: Boolean = false) {
```
Would this eliminate the need for a dedicated rule class? I'd still prefer not introducing a dedicated class but using OrtResultRule. However, if this is the only way to unblock this pull request then I will do it.  So if you insist on this dedicated class, maybe you have a good idea for a name? (I don't)

Comment (fviernau) [reply cli/src/main/kotlin/commands/EvaluatorCommand.kt]: Dropped this.

File: evaluator/src/main/kotlin/Evaluator.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/Evaluator.kt]: I changed things so that it fits what I understand you have in mind. I believe all should be covered now, except for the following two things for which I was lacking your input, so I did my best guess as follows:

1. I used injection via setter override, as to my understanding you seem to dislike injecting the resolver into the constructor or the corresponding factory method. Please see if you like that more.
2. I used the name `OrtResultRule2`. Please propose a better name as you pushed for introducing this class.

Comment (sschuberth) [reply evaluator/src/main/kotlin/Evaluator.kt]: > So, can you tell which way of injecting a directory from the test would be acceptable for you? (Ideally along with an argumentation why)

If there was a rule that inherits from `OrtResultRule` and that takes a `projectSourceTree` as a constructor argument, wouldn't that work to run the tests on that new rule, passing in "fake" sources to run the test on?

> Note that I understood from previous talks that you didn't like the construct, however I didn't get why

My main arguments for not passing a `projectSourceTree` to the `Evaluator`'s constructor are:
- Not every rule needs this, and evaluating it / cloning the project source code (accidentally) is expensive.
- The information for a `SourceTree` can be derived from the `ortResult` which already gets passed. IMO there is no need (and it's bad design) to extract some more or less arbitrary bit of information from the `ortResult` and pass a class building upon it independently.

Comment (fviernau) [reply evaluator/src/main/kotlin/Evaluator.kt]: > Passing a `SourceTree` to the constructor of `OrtResultRule2` would be of course fine! Doesn't that work?

Hm, I see. In the previous versions of my change it was a problem that this `SourceTreeResolver` was part of the user facing API. Did I just misunderstand that? (No worries, I'll change the code now to pass it into the constructor)

> I believe the proposal I had mentioned in our call earlier this week was ProjectSourceRule.

Let me just rename it quickly.

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/ProjectSourceRule.kt]: This should be `@Suppress("unused")` instead. But why are you introducing the property in this commit already if it's unused?

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:35]: Nit: Static content like `companion object`s should go first.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:36]: Formatting: The function body should start in the next line.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:47]: Maybe wrap this like

            createOrtTempDir().also { downloadDir ->
                val downloader = Downloader(DownloaderConfiguration())
                val pkg = Package.EMPTY.copy(vcsProcessed = vcsInfo)

                downloader.download(pkg, downloadDir)
            }

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/ProjectSourceRule.kt:40]: The explicit type declaration could be dropped.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:50]: Nit: As you use "Repository" instead of "Repo" in the function name above, maybe also spell out "Directory" as part of the function name here?

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (sschuberth) [reply evaluator/src/main/kotlin/ProjectSourceRule.kt]: BTW, even with the later commits, it looks like this property indeed can be private.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (sschuberth) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:38]: Nit: I'd replace the ":" with "from" and not quote the URL to align with our logging style.

**Review by sschuberth — DISMISSED** [review]

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/ProjectSourceRule.kt]: The property is public intentionally, so that user can used it to write custom matchers.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/SourceTreeResolver.kt:47]: I'll keep it as is.

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/ProjectSourceRule.kt:34]: As this is public, I kept the more specific name and solved the mentioned issue in an alternative way.

Comment (sschuberth) [reply evaluator/src/main/kotlin/ProjectSourceRule.kt]: > As this is public, I kept the more specific name

I'm not getting that argument. `projectSourceTreeResolver` is a constructor argument, not a public property. And it's in newly added code. So renaming `projectSourceTreeResolver` to `resolver` would not break the experience for any user.

> and solved the mentioned issue in an alternative way.

Personally, I don't like the "trick" with the private extension function. IMO this additional code makes it harder to read, as you need to look at more places. The idea was to reduce the code / use fewer "characters", not more.

**Comment by sschuberth** [comment]
As the rules expert, I'd like @mnonnenmacher to approve.

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/ProjectSourceRule.kt:33]: What about `sourceTreeResolver` (like the type) or `projectSourceRule` (like the rule class) to make it shorter?

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/ProjectSourceRule.kt:60]: Just `createSourceTreeResolver` would be fine IMO in this context.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:38]: The `Downloader` already prints a very similar log:
https://github.com/oss-review-toolkit/ort/blob/2d27a2244f660e43b192058221805721033aa7ce/downloader/src/main/kotlin/Downloader.kt#L187-L189

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/SourceTreeResolver.kt:41]: You should set `DownloaderConfiguration.sourceCodeOrigin = listOf(SourceCodeOrigin.VCS)` to be explicit, otherwise if the download fails there will be additional logs about the failed source artifact download which could be confusing as projects never have a source artifact.

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/ProjectSourceRule.kt:44]: Is this function supposed to be public?

File: evaluator/src/test/kotlin/ProjectSourceRuleTest.kt
Comment (mnonnenmacher) [inline evaluator/src/test/kotlin/ProjectSourceRuleTest.kt:38]: Nit: It would actually be shorter and less nested to just write:
```kotlin
val dir = createSpecTempDir()
dir.addFiles(
  "README.md",
   "module/docs/LICENSE.txt"
)
```

Or if `addFiles` would return the directory it could just be:
```kotlin
val dir = createSpecTempDir().addFiles(
  "README.md",
  "module/docs/LICENSE.txt"
)
```

(same for `addDirs`)

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (mnonnenmacher) [inline evaluator/src/main/kotlin/ProjectSourceRule.kt:50]: This should say "at least one **file**".

File: evaluator/src/test/kotlin/ProjectSourceRuleTest.kt
Comment (mnonnenmacher) [inline evaluator/src/test/kotlin/ProjectSourceRuleTest.kt:71]: Nit: Could be a one-liner.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/ProjectSourceRule.kt]: > I'm not getting that argument. `projectSourceTreeResolver` is a constructor argument, not a public property. And it's in newly added code. So renaming `projectSourceTreeResolver` to `resolver` would not break the experience for any user.

My bad, I accidentally thought it was a public property as it was one before the refactoring requested by you.

> Personally, I don't like the "trick" with the private extension function. IMO this additional code makes it harder to read, as you need to look at more places. The idea was to reduce the code / use fewer "characters", not more.

File: evaluator/src/main/kotlin/SourceTreeResolver.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/SourceTreeResolver.kt:38]: I thought it was good to at least print one log statement when doing something exotic like downloading the sources from the evaluator. If you agree, maybe we can re-phrase the log comment like e.g. 

"The evaluator rules require the project source code. Cloning $url."

What do you think?

File: evaluator/src/main/kotlin/ProjectSourceRule.kt
Comment (fviernau) [reply evaluator/src/main/kotlin/ProjectSourceRule.kt:44]: Yes: The use cases are

1. Use it from the policy rule directly (instead of the matcher)
2. Write custom matchers, re-using logic from ORT. Not that any code implemented in `Matcher.matches()` is 
    not really re-usable, so I minimize it. 

For 1, note that when implementing a policy rule often times using the matchers with their binary result is ok. But often times when one wants to write good error messages or how-to-fix, the binary result is not sufficient anymore.

File: evaluator/src/test/kotlin/ProjectSourceRuleTest.kt
Comment (fviernau) [reply evaluator/src/test/kotlin/ProjectSourceRuleTest.kt:38]: Are you ok with keeping this as is. In a following PR further tests will be added which call both `addFiles()` and `addDirs()`, so it will be consistent.

If you still prefer to change it, can I do this in a follow up PR, to avoid the rebase hazzle with my following changes?

Comment (mnonnenmacher) [reply evaluator/src/test/kotlin/ProjectSourceRuleTest.kt:38]: Fleunt API could still be useful but I'll leave that up to you.

Comment (fviernau) [reply evaluator/src/test/kotlin/ProjectSourceRuleTest.kt:38]: I'll keep it in mind and probably implement it in a following PR.

**Review by mnonnenmacher — APPROVED** [review]

**Pull request merged**
